File: src/dpl_pipeline/rules.py
"""Rule loading and matching.

Rules live in a YAML lexicon (see rules/impact_rules.yml). Each rule is a
small, auditable pattern set:

    - id: R-ID-001
      category: restrictive
      policy_area: voter_id
      description: New or stricter photo ID requirement for in-person voting
      any:   [list of regexes -- at least one must match]
      all:   [optional list -- every one must also match]
      none:  [optional list -- rule is suppressed if any of these match]

Matching is pure and order-independent per rule; rules are always evaluated
in sorted-id order so results are deterministic regardless of YAML ordering.
All patterns are compiled case-insensitively with DOTALL so `.{0,80}` style
proximity windows can cross line breaks.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path

import yaml

from .models import CATEGORIES

_FLAGS = re.IGNORECASE | re.DOTALL


@dataclass(frozen=True)
class Rule:
    rule_id: str
    category: str
    policy_area: str
    description: str
    any_patterns: tuple[re.Pattern, ...]
    all_patterns: tuple[re.Pattern, ...]
    none_patterns: tuple[re.Pattern, ...]

    def match(self, text: str) -> re.Match | None:
        """Return the first `any` match if the rule fires on `text`, else None."""
        for p in self.none_patterns:
            if p.search(text):
                return None
        for p in self.all_patterns:
            if not p.search(text):
                return None
        first: re.Match | None = None
        for p in self.any_patterns:
            m = p.search(text)
            if m and (first is None or m.start() < first.start()):
                first = m
        return first


class RulesError(ValueError):
    pass


def _compile(patterns: list[str], rule_id: str, key: str) -> tuple[re.Pattern, ...]:
    out = []
    for pat in patterns:
        try:
            out.append(re.compile(pat, _FLAGS))
        except re.error as e:
            raise RulesError(f"rule {rule_id}: bad regex in `{key}`: {pat!r} ({e})") from e
    return tuple(out)


def load_rules(path: str | Path) -> list[Rule]:
    """Load, validate, and compile the rule lexicon. Returns rules sorted by id."""
    raw = yaml.safe_load(Path(path).read_text(encoding="utf-8"))
    if not isinstance(raw, list) or not raw:
        raise RulesError(f"{path}: expected a non-empty list of rules")

    rules: list[Rule] = []
    seen: set[str] = set()
    for i, item in enumerate(raw):
        rid = item.get("id")
        if not rid:
            raise RulesError(f"{path}: rule #{i} has no id")
        if rid in seen:
            raise RulesError(f"{path}: duplicate rule id {rid}")
        seen.add(rid)
        cat = item.get("category")
        if cat not in CATEGORIES:
            raise RulesError(f"rule {rid}: category must be one of {CATEGORIES}, got {cat!r}")
        if not item.get("policy_area"):
            raise RulesError(f"rule {rid}: policy_area is required")
        any_pats = item.get("any") or []
        if not any_pats:
            raise RulesError(f"rule {rid}: `any` must contain at least one pattern")
        rules.append(
            Rule(
                rule_id=rid,
                category=cat,
                policy_area=item["policy_area"],
                description=item.get("description", ""),
                any_patterns=_compile(any_pats, rid, "any"),
                all_patterns=_compile(item.get("all") or [], rid, "all"),
                none_patterns=_compile(item.get("none") or [], rid, "none"),
            )
        )
    rules.sort(key=lambda r: r.rule_id)
    return rules


def default_rules_path() -> Path:
    """rules/impact_rules.yml at the repository root (two levels up from this file)."""
    return Path(__file__).resolve().parents[2] / "rules" / "impact_rules.yml"

File: lib/rules.ts
/**
 * Rule loading and matching.
 *
 * Rules live in a YAML lexicon (rules/impact_rules.yml). Each rule:
 *
 *   - id: R-ID-001
 *     category: restrictive
 *     policy_area: voter_id
 *     description: ...
 *     any:  [regexes — at least one must match]        (required)
 *     all:  [regexes — every one must also match]      (optional)
 *     none: [regexes — rule suppressed if any matches] (optional)
 *
 * Patterns are compiled case-insensitively with dotAll so `.{0,80}` style
 * proximity windows can cross line breaks. Rules are always evaluated in
 * sorted-id order so results are deterministic regardless of YAML order.
 */

import { readFileSync } from "node:fs";
import path from "node:path";
import { load } from "js-yaml";

import { CATEGORIES, type Category } from "./types";

export interface Rule {
  id: string;
  category: Category;
  policyArea: string;
  description: string;
  any: RegExp[];
  all: RegExp[];
  none: RegExp[];
}

export class RulesError extends Error {}

interface RawRule {
  id?: string;
  category?: string;
  policy_area?: string;
  description?: string;
  any?: string[];
  all?: string[];
  none?: string[];
}

function compile(patterns: string[], ruleId: string, key: string): RegExp[] {
  return patterns.map((p) => {
    try {
      return new RegExp(p, "is");
    } catch (e) {
      throw new RulesError(`rule ${ruleId}: bad regex in \`${key}\`: ${p} (${e})`);
    }
  });
}

export function parseRules(yamlText: string, source = "rules"): Rule[] {
  const raw = load(yamlText);
  if (!Array.isArray(raw) || raw.length === 0) {
    throw new RulesError(`${source}: expected a non-empty list of rules`);
  }
  const seen = new Set<string>();
  const rules = (raw as RawRule[]).map((item, i) => {
    const id = item.id;
    if (!id) throw new RulesError(`${source}: rule #${i} has no id`);
    if (seen.has(id)) throw new RulesError(`${source}: duplicate rule id ${id}`);
    seen.add(id);
    if (!CATEGORIES.includes(item.category as Category)) {
      throw new RulesError(`rule ${id}: category must be one of ${CATEGORIES.join(", ")}, got ${item.category}`);
    }
    if (!item.policy_area) throw new RulesError(`rule ${id}: policy_area is required`);
    if (!item.any?.length) throw new RulesError(`rule ${id}: \`any\` must contain at least one pattern`);
    return {
      id,
      category: item.category as Category,
      policyArea: item.policy_area,
      description: item.description ?? "",
      any: compile(item.any, id, "any"),
      all: compile(item.all ?? [], id, "all"),
      none: compile(item.none ?? [], id, "none"),
    };
  });
  rules.sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0));
  return rules;
}

export function defaultRulesPath(): string {
  return path.join(process.cwd(), "rules", "impact_rules.yml");
}

let cached: Rule[] | null = null;

/** Load the default lexicon once per server process. */
export function loadRules(filePath?: string): Rule[] {
  if (filePath) return parseRules(readFileSync(filePath, "utf-8"), filePath);
  if (!cached) cached = parseRules(readFileSync(defaultRulesPath(), "utf-8"), defaultRulesPath());
  return cached;
}

/**
 * If the rule fires on `text`, return the earliest `any` match; else null.
 */
export function matchRule(rule: Rule, text: string): RegExpExecArray | null {
  for (const p of rule.none) if (p.test(text)) return null;
  for (const p of rule.all) if (!p.test(text)) return null;
  let first: RegExpExecArray | null = null;
  for (const p of rule.any) {
    const m = p.exec(text);
    if (m && (first === null || m.index < first.index)) first = m;
  }
  return first;
}
